utils/sorting.hh

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <stdexcept>
+#include <vector>
+#include <map>
+#include <seastar/core/future.hh>
+#include "seastar/coroutine/maybe_yield.hh"
+
+namespace utils {
+
+/**
+ * Topological sort a DAG using Kahn's algorithm.
+ *
+ * @param vertices      Contains all vertices of the graph
+ * @param adjecency_map Entry {k, v} means there is an k->v edge in the graph
+ * @return a vector of topologically sorted vertices
+ * @throws std::runtime_error when a graph has any cycle.
+ */
+template<typename T, typename Compare = std::less<T>>
+seastar::future<std::vector<T>> topological_sort(const std::vector<T>& vertices, const std::multimap<T, T, Compare>& adjecency_map) {
+    std::map<T, uint32_t, Compare> ref_count_map; // Contains counters how many edges point (referece) to a vertex
+    std::vector<T> sorted;
+    sorted.reserve(vertices.size());
+
+    for (auto& edge: adjecency_map) {
+        ref_count_map[edge.second]++;
+    }
+
+    // Push to result vertices which reference counter == 0
+    for (auto& v: vertices) {
+        if (!ref_count_map.contains(v)) {
+            sorted.emplace_back(v);
+        }
+    }
+
+    // Iterate over already sorted vertices, 
+    // decrease counter of vertices to which sorted vertices points,
+    // push to the end of `sorted` vector vertices which counter == 0.
+    for (size_t i = 0; i < sorted.size(); i++) {
+        auto [it_begin, it_end] = adjecency_map.equal_range(sorted[i]);
+        
+        while (it_begin != it_end) {
+            auto d = it_begin->second;
+            if (--ref_count_map[d] == 0) {
+                sorted.push_back(d);
+            }
+            ++it_begin;
+        }
+        co_await seastar::coroutine::maybe_yield();
+    }
+
```

Comment:
Suggestion: use `utils::clear_gently` to clear the `ref_count_map`. This reduces the chance of a stall.